LaTeX/LaTeX.sublime-syntax

```diff
@@ -198,6 +208,134 @@ contexts:
+
+  xparse-single-argspec:
+    # modifiers and simple specs
+    - match: '[+!]'
+      scope: storage.modifier.latex
+    - match: '[movsb]'
+      scope: storage.type.latex
+    - match: t
+      scope: storage.type.latex
+      push: xparse-args-token
+
+    # delimited arguments without default
+    - match: '[rd]'
+      scope: storage.type.latex
+      push:
+        - xparse-args-token
+        - xparse-args-token
+
+    # delimited arguments with default
+    - match: '[RD]'
+      scope: storage.type.latex
+      push:
+        - xparse-optarg-value-meta
+        - xparse-args-token
+        - xparse-args-token
+
+    # optional argument with default
+    - match: O
+      scope: storage.type.latex
+      push: xparse-optarg-value-meta
+
+    # embellishment argument
+    - match: e
+      scope: storage.type.latex
+      push: xparse-embellishment
+
+    # embellishment argument with default
+    - match: E
+      scope: storage.type.latex
+      push:
+        - xparse-optarg-value-meta
+        - xparse-embellishment
+
+  xparse-embellishment:
+    - match: \{
+      scope: punctuation.definition.group.brace.begin.latex
+      set: xparse-embellishment-content
+    - match: (?=\S)
+      set: xparse-args-token
+    - include: paragraph-pop
+
+  xparse-embellishment-content:
+    - match: (?!\})
+      push: xparse-args-token
+    - include: brace-group-end
+    - include: paragraph-pop
+    - include: else-pop
+
+  xparse-args-token:
+    # either a command
+    - match: (\\){{cmdname}}{{endcs}}
+      scope: constant.other.token.latex
+      captures:
+        1: punctuation.definition.backslash.latex
+      pop: 1
+    # or the token wrapped in a group
+    - match: \{
+      scope: punctuation.definition.group.brace.begin.tex
+      push: xparse-args-token
+    - match: \}
+      scope: punctuation.definition.group.brace.end.tex
+      pop: 1
+    # or a single character
+    - match: '\S'
+      scope: constant.character.latex constant.other.token.latex
+      pop: 1
+    - include: paragraph-pop
+    - include: else-pop
+
+  xparse-optarg-value-meta:
+    - meta_include_prototype: false
+    - match: ''
+      set: xparse-optarg-value
+
+  xparse-optarg-value:
+    - clear_scopes: 1
+    - meta_scope: meta.function.parameters.default-value.latex
+    - match: \{
+      scope: punctuation.definition.group.brace.begin.latex
+      push: xparse-optarg-brace-body
+    - match: (\\){{cmdname}}{{endcs}}
+      scope:  support.function.latex
+      captures:
+        1: punctuation.definition.backslash.latex
+    - match: (?=\})
+      pop: 1
+    - match: '\S'
+      pop: 1
+    - include: paragraph-pop
+
+  xparse-optarg-brace-body:
+    - match: \}
+      scope: punctuation.definition.group.brace.end.latex
+      pop: 2
+    - include: general-constants
+    - include: general-commands
+    - include: macro-braces
+    # The optional value is a long argument (can span paragraphs), so no
+    # paragraph-pop here!
+
```

Comment:
I'd try to clearly distinguish braced and unbraced value types by using two different contexts to implement them. It may be more robust, avoids cross references and is therefore more flexible with regards two which constructs are supported in each.

```suggestion
  xparse-optarg-value:
    - meta_include_prototype: false
    - match: (?={)
      set:
        - xparse-optarg-group-body
        - brace-group-begin
    - match: (?=\S)
      set: xparse-optarg-value-body
    - include: paragraph-pop

  xparse-optarg-group-body:
    - clear_scopes: 1
    - meta_scope: meta.function.parameters.default-value.latex meta.group.brace.latex
    - include: brace-group-end
    - include: general-constants
    - include: general-commands
    - include: macro-braces
    # The optional value is a long argument (can span paragraphs), so no
    # paragraph-pop here!

  xparse-optarg-value-body:
    - clear_scopes: 1
    - meta_scope: meta.function.parameters.default-value.latex
    - match: (\\){{cmdname}}{{endcs}}
      scope: support.function.latex
      captures:
        1: punctuation.definition.backslash.latex
    - match: (?=\})
      pop: 1
    - match: \S
      pop: 1
    - include: paragraph-pop

```

I am a bit uncertain about unbraced default values as I can't find any statements about what's supported here. This context looks a bit suspicious to me, because `support.function` pattern doesn't pop, while a single non-whitespace `\S` does after being consumed.

How about `macro-placeholders`? They seem to be supported within default values.

```latex
\NewDocumentCommand {\margins} { O{#3} m O{#1} m } {(#1,#2,#3,#4)}
```

Reply:
yes, macro placeholders are supported. I think the single command should also pop. If there are no braces, then TeX reads the next token, which is either the next non-whitespace character or the command sequence.